### PR TITLE
Implement more CoAP Options

### PR DESCRIPTION
**Pull request by GiedriusM**
* Observe
* Max-Age
* Uri-Query

### CODE REVIEW

**Comment by GiedriusM** [comment]
I used these in my application, might be useful for others too.

**Comment by GiedriusM** [comment]
Also, I've noticed that you must call option append functions in order by their CoAP code.
This works:
`
otCoapHeaderAppendObserveOption(&reqHeader, 0);
otCoapHeaderAppendUriPathOptions(&reqHeader, "path");
otCoapHeaderAppendMaxAgeOption(&reqHeader, 1234);
otCoapHeaderAppendUriQueryOption(&reqHeader, "key=value");
`

This doesn't:
`
otCoapHeaderAppendUriPathOptions(&reqHeader, "path");
otCoapHeaderAppendUriQueryOption(&reqHeader, "key=value");
otCoapHeaderAppendObserveOption(&reqHeader, 0);
otCoapHeaderAppendMaxAgeOption(&reqHeader, 1234);
`
Observe and Max-Age are not added because kCoapOptionObserve, kCoapOptionMaxAge < kCoapOptionUriQuery.

Should I create a [known-]issue for this?

File: src/core/coap/coap_header.cpp [hunk 1]
```diff
@@ -244,6 +244,32 @@ ThreadError Header::AppendOption(const Option &aOption)
...
 }
 
+ThreadError Header::AppendObserveOption(uint32_t aObserve)
+{
+    Option coapOption;
+    uint8_t buf[3];
+
+    coapOption.mNumber = kCoapOptionObserve;
+
+    if (aObserve <= 0xff) {
```
Comment (aeliot) [inline src/core/coap/coap_header.cpp]: For `if` statements, the `{` should be on its own line.
```
if (...)
{
    ....
}
else
{
    ....
}
```

File: src/core/coap/coap_header.cpp [hunk 2]
```diff
@@ -281,6 +307,49 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
...
 }
 
+ThreadError Header::AppendMaxAgeOption(uint32_t aMaxAge)
+{
+    Option coapOption;
+    uint8_t buf[4];
+
+    coapOption.mNumber = kCoapOptionMaxAge;
+
+    if (aMaxAge <= 0xff) {
```
Comment (aeliot) [inline src/core/coap/coap_header.cpp]: Same note as above, curly braces should have their own lines.

File: src/core/coap/coap_header.cpp [hunk 3]
```diff
@@ -281,6 +307,49 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
...
+    return AppendOption(coapOption);
+}
+
+ThreadError Header::AppendUriQueryOption(const char *aUriQuery)
+{
+    Option coapOption;
+
+    coapOption.mNumber = kCoapOptionUriQuery;
+    coapOption.mLength = strlen(aUriQuery);
+    coapOption.mValue = (uint8_t *)aUriQuery;
```
Comment (aeliot) [inline src/core/coap/coap_header.cpp]: Can you align the `=` here?

File: src/core/coap/coap_header.cpp [hunk 4]
```diff
@@ -281,6 +312,56 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
...
+    return AppendOption(coapOption);
+}
+
+ThreadError Header::AppendUriQueryOption(const char *aUriQuery)
+{
+    Option coapOption;
+
+    coapOption.mNumber = kCoapOptionUriQuery;
+    coapOption.mLength = static_cast<uint16_t>(strlen(aUriQuery));
+    coapOption.mValue  = reinterpret_cast<const uint8_t *>(aUriQuery);
```
Comment (rlubos) [inline src/core/coap/coap_header.cpp]: Is reinterpret_cast really needed here, wouldn't static_cast be enough?

File: src/core/coap/coap_header.cpp [hunk 5]
```diff
@@ -281,6 +312,56 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
...
 }
 
+ThreadError Header::AppendMaxAgeOption(uint32_t aMaxAge)
+{
+    Option coapOption;
+    uint8_t buf[4];
+
+    coapOption.mNumber = kCoapOptionMaxAge;
+
+    if (aMaxAge <= 0xff)
```
Comment (rlubos) [inline src/core/coap/coap_header.cpp]: Have you considered using a loop to make the code more concise? You could make it like:
```
aMaxAge = Encoding::BigEndian::HostSwap32(aMaxAge);
coapOption.mLength = 0;
do
{
    buf[coapOption.mLength] = aMaxAge & 0xff;
    aMaxAge = aMaxAge >> 8;
    coapOption.mLength++;
} while (aMaxAge != 0 && coapOption.mLength < sizeof(buf))
```
You could even make an generic function from it and use in both, AppendMaxAgeOption and AppendObserveOption.

File: include/openthread-coap.h [hunk 6]
```diff
@@ -224,6 +239,31 @@ ThreadError otCoapHeaderAppendOption(otCoapHeader *aHeader, const otCoapOption *
...
+ * @retval kThreadError_NoBufs       The option length exceeds the buffer size.
+ */
+ThreadError otCoapHeaderAppendMaxAgeOption(otCoapHeader *aHeader, uint32_t aMaxAge);
+
+/**
+ * This function appends a single Uri-Query option.
+ *
+ * @param[inout]  aHeader   A pointer to the CoAP header.
+ * @param[in]     aUriQuery A pointer to NULL-terminated string, which should contain
+ *  a single key=value pair.
```
Comment (GiedriusM) [inline include/openthread-coap.h]: I didn't want it to be on the same column as @param, so just let the IDE to auto-align it. I'll make it one line, seeing as there are longer lines in that file.

File: src/core/coap/coap_header.cpp (hunk 5 above)
Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: Will do

File: src/core/coap/coap_header.cpp (hunk 4 above)
Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: I was planning on doing static cast, but noticed that AppendUriPathOptions uses reinterpret and just went with common "style".

File: include/openthread-coap.h (hunk 6 above)
Comment (rlubos) [inline include/openthread-coap.h]: By alignment I meant to align it to a column where description begins (some other docs do it this way):
```
* @param[in]     aUriQuery A pointer to NULL-terminated string, which should contain
*                          a single key=value pair.
```
But a single line should work as well 👍

File: src/core/coap/coap_header.cpp (hunk 4 above)
Comment (rlubos) [inline src/core/coap/coap_header.cpp]: So it must've been missed at some point, as generally static_cast's are preferred. There was even a PR some time ago to get rid of reinterpret_cast wherever possible :)

Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: Do you want me to replace  reinterpret to static cast in AppendUriPathOptions? Or keep the focus only on the new code?

Comment (jwhui) [inline src/core/coap/coap_header.cpp]: Go ahead and replace with `static_cast()`.  The change is small and contained within the same file.  Thanks!

File: src/core/coap/coap_header.cpp (hunk 5 above)
Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: On a second thought, are you sure this code would work? E.g. given aMaxAge = 0x00123456:
on LE machines: aMaxAge => 0x56341200, buf = { 00, 12, 34, 56 } - good order, length too long
on BE machines: aMaxAge => 0x00123456, buf = { 56, 34, 12 } - reversed order, good length.
Or am I missing something?
I think this should give the correct result:
```C++
aMaxAge = Encoding::BigEndian::HostSwap32(aMaxAge);
coapOption.mLength = 4;
coapOption.mValue = static_cast<uint8_t *>(&aMaxAge);
// skip preceding zeros, but make sure mLength is at least 1
while (coapOption.mValue[0] == 0 && coapOption.mLength > 1) {
    coapOption.mValue++;
    coapOption.mLength--;
}
```
, but not so sure if short and vague is better than long and clear.

File: src/core/coap/coap_header.cpp (hunk 4 above)
Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: Hah, that's the reason for reinterpret_cast:
> error: invalid static_cast from type 'const char*' to type 'const uint8_t* {aka const unsigned char*}'

Looking at http://eel.is/c++draft/basic.types#basic.fundamental probably explains why

> Plain char, signed char, and unsigned char are **three distinct types**, collectively called narrow character types

Comment (jwhui) [inline src/core/coap/coap_header.cpp]: Yes, I remember now.  That's why I had it as `reinterpret_cast` to begin with :)

File: src/core/coap/coap_header.cpp (hunk 5 above)
Comment (GiedriusM) [inline src/core/coap/coap_header.cpp]: Alright, I'll try to test and push the changes tomorrow.

Comment (rlubos) [inline src/core/coap/coap_header.cpp]: Hey, you're right, I've took it too fast with the example and it had issues you've pointed, sorry about that! I really like the solution you've proposed, especially it seems quite generic and could be reused in both options.

**Comment by GiedriusM** [comment]
Sorry I didn't have time to do the changes yesterday, as I said I would. I'm pushing them now. Fixed comment alignment and changed code to be more concise. Unit tested it locally on my machine, seemed to work. Currently don't have access to real setup to sniff if everything is good over-the-air, but I'm pretty sure it will be. Please check if I haven't missed anything.

**Comment by GiedriusM** [comment]
@jwhui P.S. this solution saved 38 bytes for Cortex M0+ processor :)

**Comment by jwhui** [comment]
@GiedriusM, thanks for the update!

BTWk looks like clang static analyzer is failing on you latest update:
```
coap/coap_header.cpp:186:44: warning: The left operand of '-' is a garbage value
    uint16_t optionDelta = aOption.mNumber - mOptionLast;
```

Also, make sure to run `astyle` before pushing your changes.

**Comment by GiedriusM** [comment]
Left an embarrassing mistake. Rushing, while switching between different repos and environments does not help :) Hopefully, it's all good now. Tested in over the air setup.

**Pull request merged**